src/realm/util/sensitive_buffer.cpp

```diff
@@ -0,0 +1,227 @@
+        // Since we don't use VirtualAlloc with real page size and assign buffer to the portion
+        // of the allocated page (this'd complicate the logic significantly),
+        // simply try to double the limit a few times and attempt to lock the buffer again
+        // in case when multiple threads do the same. The min limit itself is pretty small,
+        // and the doc says that even this is not strictly guaranteed by the system, so
+        // there should be no harm in overcommiting expected min working set size
+        // In real tests: even with 128 threads on 16 cores system in 32bit app this loop
+        // may go through 4-5 iterations until the lock succeeds if the increment
+        // is just a multiple of a few pages, but with doubling of the limit
+        // it should succeed on the first try.
+        min_size *= 2;
+        max_size = std::max(4 * min_size, max_size);
+
+        // give up and don't try to ask the system to keep more than 90% memory resident in the process
+        if (max_size > 0.9 * limits->mem_size)
+            break;
+
+        BOOL ret = SetProcessWorkingSetSize(GetCurrentProcess(), min_size, max_size);
+        REALM_ASSERT_RELEASE_EX(ret != 0 && "SetProcessWorkingSetSizeEx", GetLastError());
+        limits->min_size = min_size;
+        limits->max_size = max_size;
+    }
+
+    // the loop above didn't succeed, provide some mem info and assert
+    MEMORYSTATUSEX mem;
+    mem.dwLength = sizeof(mem);
+    GlobalMemoryStatusEx(&mem);
+    REALM_ASSERT_RELEASE_EX(res != 0 && "VirtualLock()", err, mem.ullAvailPhys, mem.ullTotalPhys, mem.dwMemoryLoad,
+                            mem.ullAvailPageFile, mem.ullAvailVirtual);
+}
+
+SensitiveBufferBase::SensitiveBufferBase(size_t size)
+    : m_size(size_t(ceil((long double)size / CRYPTPROTECTMEMORY_BLOCK_SIZE) * CRYPTPROTECTMEMORY_BLOCK_SIZE))
+{
+    m_buffer = VirtualAlloc(nullptr, m_size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+    REALM_ASSERT_RELEASE_EX(m_buffer != NULL && "VirtualAlloc()", GetLastError());
+
+    lock_or_grow_working_size(m_buffer, m_size);
+}
+
+SensitiveBufferBase::~SensitiveBufferBase()
+{
+    if (!m_buffer)
+        return;
+
+    secure_erase(m_buffer, m_size);
+
+    BOOL ret = VirtualUnlock(m_buffer, m_size);
+    REALM_ASSERT_RELEASE_EX((ret != 0 || GetLastError() == ERROR_NOT_LOCKED) && "VirtualUnlock()", GetLastError());
+
+    ret = VirtualFree(m_buffer, 0, MEM_RELEASE);
+    REALM_ASSERT_RELEASE_EX(ret != 0 && "VirtualFree()", GetLastError());
+
+    m_buffer = nullptr;
+}
+
+void SensitiveBufferBase::protect() const
+{
+    // MEMO even though we try to lock the page with the buffer, and that should prevent it to be swapped,
+    //      locking is not reliable and may fail under high demand in the system due to some opaque reason,
+    //      which we can't recover from, so use second layer to protect the buffer if it is swapped
+    //      note: look at attempt_to_lock where it's expected
```

Comment:
We lock the buffer's address space into physical memory to prevent its contents from leaking in the swapfile. We additionally encrypt it to prevent it leaking from a dump of the process's memory. Just locking is not enough to guard against the second scenario. But encrypting it might also be sufficient against the first?

Reply:
i'm not sure about this. The mechanism of inner workings of this function is not clearly documented. I'd assume that the key for this function is stored automatically in some locked page, so that VirtualLock/Unlock is not really needed at all. But the key then may be swapped out still. And there is no info if it's some global key, or per process.

The doc page explicitly say that:
"Using [CryptProtectMemory](https://learn.microsoft.com/en-us/windows/desktop/api/dpapi/nf-dpapi-cryptprotectmemory) and [CryptUnprotectMemory](https://learn.microsoft.com/en-us/windows/desktop/api/dpapi/nf-dpapi-cryptunprotectmemory) for password encryption is not secure because the data exists as plaintext in memory before it is encrypted and at any time the caller decrypts it for use."
So i'm not sure about this whole approach without page lock. When the buffer is decrypted, it can be swapped or memory can be dumped. 
On the other hand, from what i read in the doc VirtualLock is  just a hint, and so the OS is allowed to ignore it. So it may be not worse it after all. We can do just with CryptProtectMemory.

Reply:
AFAIK the key used by `CryptProtectMemory` lives in kernel space, so just dumping the process memory won't reveal it. The key is specific to this process because we call `CryptProtectMemory` with `CRYPTPROTECTMEMORY_SAME_PROCESS` so another process on the same machine with access to the memory won't be able to decrypt it unless the kernel is compromised.
The customer that requested this feature is aware that we regularly have to decrypt the key in the course of reading and writing from disk and that there is the chance to do a memory dump exactly when the key is decrypted. But while we can't shield the key perfectly, we can improve on the baseline, and a combination of `VirtualLock` and `CryptProtectMemory` can diminish the effectiveness of a rudimentary attack, which is all we can do with the tools that we have. It's on the customer to further protect their data by potentially detecting and refusing to run with a debugger attached or obfuscating their executable code, but that's beyond our scope.
`CryptProtectMemory` and friends can be a footgun if you don't understand their pitfalls and let them lull you in a false sense of security, but our assumption here is that using them is like locking your front door - a determined thief can pick the lock, but an opportunistic thief will be deterred, i.e. "security" is more a gradient than a binary condition.